### PR TITLE
planner: complete fall through into MPP task if up-layer can generate a cop tiflash 

### PR DESCRIPTION
**Pull request by guo-shaoge**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/37459

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/find_best_task.go [hunk 1]
```diff
@@ -2181,30 +2180,18 @@ func (ds *DataSource) convertToTableScan(prop *property.PhysicalProperty, candid
...
-	if prop.TaskTp == property.MppTaskType || canMppConvertToRootForDisaggregatedTiFlash {
+	isTiFlashPath := ts.StoreType == kv.TiFlash
+	canMppConvertToRoot := prop.TaskTp == property.RootTaskType && ds.SCtx().GetSessionVars().IsMPPAllowed()
+	if prop.TaskTp == property.MppTaskType || canMppConvertToRoot {
 		if ts.KeepOrder {
 			return invalidTask, nil
 		}
-		if prop.MPPPartitionTp != property.AnyType || (ts.isPartition && !isDisaggregatedTiFlash) {
+		if prop.MPPPartitionTp != property.AnyType {
 			// If ts is a single partition, then this partition table is in static-only prune, then we should not choose mpp execution.
```
Comment (guo-shaoge) [inline planner/core/find_best_task.go:2190]: 1. Why this if branch.
2. Del useless comment.

File: sessionctx/variable/tidb_vars.go [hunk 2]
```diff
@@ -373,6 +373,11 @@ const (
 	// Value set to `false` means never use mpp.
 	TiDBAllowMPPExecution = "tidb_allow_mpp"
 
+	// TiDBBanTiFlashCop means we only use MPP mode to query data.
+	// Default value is `true`, means to be determined by the optimizer.
+	// Value set to `false` means we may fall back to TiFlash cop plan if possible.
+	TiDBBanTiFlashCop = "tidb_ban_tiflash_cop"
```
Comment (fixdb) [inline sessionctx/variable/tidb_vars.go:379]: Double negative is hard to follow. 
Can we rename to 'tidb_allow_tiflash_cop' or 'tidb_enable_tiflash_cop'?

Comment (AilinKid) [reply sessionctx/variable/tidb_vars.go:379]: make sense

**Review by fixdb — APPROVED** [review]
+1

**Review by elsa0520 — APPROVED** [review]
LGTM

File: sessionctx/variable/tidb_vars.go [hunk 3]
```diff
@@ -373,6 +373,11 @@ const (
 	// Value set to `false` means never use mpp.
 	TiDBAllowMPPExecution = "tidb_allow_mpp"
 
+	// TiDBAllowTiFlashCop means we only use MPP mode to query data.
```
Comment (guo-shaoge) [inline sessionctx/variable/tidb_vars.go:376]:
```suggestion
	// TiDBAllowTiFlashCop determines whether we can use Cop/BatchCop for tiflash query.
```

File: expression/integration_test/integration_test.go [hunk 4]
```diff
@@ -3782,13 +3782,10 @@ func TestShardIndexOnTiFlash(t *testing.T) {
...
-	for _, row := range rows {
-		line := fmt.Sprintf("%v", row)
-		if strings.Contains(line, "TableFullScan") {
-			require.NotContains(t, line, "mpp[tiflash]")
-		}
-	}
+	// when we isolated the read engine as 'tiflash' and banned TiDB opening allow-mpp, no suitable plan is generated.
+	_, err := tk.Exec("explain select max(b) from t")
+	require.NotNil(t, err)
+	require.Equal(t, err.Error(), "[planner:1815]Internal : Can't find a proper physical plan for this query")
```
Comment (guo-shaoge) [inline expression/integration_test/integration_test.go:3788]: I remember err msg has already refined(https://github.com/pingcap/tidb/pull/43908/files#diff-bca2046fd17f493604dd1124ab98d60de4aa21e57135324e365eb33e68b74f43), why not works here?

File: planner/core/find_best_task.go [hunk 5]
```diff
@@ -2191,19 +2191,18 @@ func (ds *DataSource) convertToTableScan(prop *property.PhysicalProperty, candid
 	}
 	// In disaggregated tiflash mode, only MPP is allowed, cop and batchCop is deprecated.
 	// So if prop.TaskTp is RootTaskType, have to use mppTask then convert to rootTask.
-	isDisaggregatedTiFlash := config.GetGlobalConfig().DisaggregatedTiFlash
-	isDisaggregatedTiFlashPath := isDisaggregatedTiFlash && ts.StoreType == kv.TiFlash
-	canMppConvertToRootForDisaggregatedTiFlash := isDisaggregatedTiFlashPath && prop.TaskTp == property.RootTaskType && ds.SCtx().GetSessionVars().IsMPPAllowed()
-	if prop.TaskTp == property.MppTaskType || canMppConvertToRootForDisaggregatedTiFlash {
+	isTiFlashPath := ts.StoreType == kv.TiFlash
+	canMppConvertToRoot := prop.TaskTp == property.RootTaskType && ds.SCtx().GetSessionVars().IsMPPAllowed() && isTiFlashPath
```
Comment (guo-shaoge) [inline planner/core/find_best_task.go:2195]: merge `canMppConvertToRootForWhenTiFlashCopIsBanned` and `canMppConvertToRootForWhenTiFlashCopIsBanned` into one?

Comment (AilinKid) [reply planner/core/find_best_task.go:2195]: not now, we don't want the allowTiflashCop-Switch to control the disaggregate behavior, sometimes we still need to fall  back to tiflash cop plan

File: expression/integration_test/integration_test.go (hunk 4 above)
Comment (AilinKid) [reply expression/integration_test/integration_test.go:3788]: make sense, I will check allowTiflashCop-Switch there like you do

Comment (AilinKid) [reply expression/integration_test/integration_test.go:3788]: as we talked about, t.invalid() is a lazy check, we couldn't tell what exactly cause the plan pruning and finally nothing generated

**Review by AilinKid — APPROVED** [review]

**Comment by wuhuizuo** [comment]
/hold 

wait for reviewing from configuration review.

**Review by qw4990 — APPROVED** [review]
